### PR TITLE
[bug#12464] Fix resizing for `ArrayBuffer` and `PriorityQueue`

### PR DESCRIPTION
**Pull request by NthPortal**
Fix `Int` overflows when ensuring the size of an `ArrayBuffer`.

Always use `ensureSize` in `ArrayBuffer.from` to check
against the VM array size limit.

Make sizing behaviour of `ArrayBuffer` consistent.

Partially addresses scala/bug#12464 (still need to address `ArrayDeque`)

### CODE REVIEW

File: src/library/scala/collection/mutable/ArrayBuffer.scala
Comment (OndrejSpanel) [inline src/library/scala/collection/mutable/ArrayBuffer.scala]: `n.toLong` is not really necessary here, is it? In other places the conversion is done before adding, to make it does not overflow. Here is is just passed. Making the conversion explicit does no harm, though, and it makes it obvious what is going on.

File: src/library/scala/collection/mutable/ArrayBuffer.scala
Comment (OndrejSpanel) [inline src/library/scala/collection/mutable/ArrayBuffer.scala:76]: Here .toLong is necessary. Perhaps a comment might be added to explain why? I am not sure if it is clean enough from the code.

File: src/library/scala/runtime/PStatics.scala
Comment (OndrejSpanel) [inline src/library/scala/runtime/PStatics.scala:18]: Is the hotspot limit documented somewhere? If it is, a link would come handy. I there any reason why `Int.MaxValue - 2` is only a comment? Why not writing `val VM_MaxArraySize = Int.MaxValue - 2` instead?

File: src/library/scala/collection/mutable/ArrayBuffer.scala
Comment (NthPortal) [reply src/library/scala/collection/mutable/ArrayBuffer.scala]: I mean, the method takes a `Long`, so I called `toLong`. I guess I forgot that the compiler would insert the conversion for me

File: src/library/scala/runtime/PStatics.scala
Comment (NthPortal) [reply src/library/scala/runtime/PStatics.scala:18]: I don't know where, if anywhere, the hotspot limit is documented. I used the exact value because I can never remember how much constant folding the compiler does, and I wanted to make sure it was a constant. I remember there being a limit to how much folding the compiler will do

File: src/library/scala/collection/mutable/ArrayBuffer.scala
Comment (NthPortal) [reply src/library/scala/collection/mutable/ArrayBuffer.scala:76]: if you don't do `toLong` on one of them first, it adds the two `Int`s together, which has the same overflow problem. to me, at least, that's obvious? it's fairly common to call `to<number_type>` on one argument before doing maths, to ensure that the operation is done using the correct type (e.g. calling `toDouble` to ensure you get floating point division rather than integer division).

File: src/library/scala/runtime/PStatics.scala
Comment (OndrejSpanel) [reply src/library/scala/runtime/PStatics.scala:18]: I have found this: https://stackoverflow.com/questions/3038392/do-java-arrays-have-a-maximum-size . It seems VM dependent, a value of `Integer.MAX_VALUE - 8` is mentioned in one of the answers: https://stackoverflow.com/a/8381338/16673

Comment (NthPortal) [reply src/library/scala/runtime/PStatics.scala:18]: I could perhaps be convinced to change it to that. I'm more inclined to just pick the lowest value that anyone can find on a VM though. it also doesn't hurt to reduce it later if need be

**Comment by SethTisue** [comment]
@NthPortal have the conversations above concluded? do you think (does @scala/collections think) this has now received sufficient review?

**Comment by SethTisue** [comment]
Haha oops, Dale just noticed that CI is failing. It's just MiMa, but of course that needs to be resolved.

**Comment by NthPortal** [comment]
I don't have strong feelings about how to resolve the conversations. I'm inclined to leave the max array size as is, until shown otherwise, but have no feelings about the other. I've been quite under the weather for over a week now, and haven't really had the energy to work on anything. if someone wants to decide one way or another on the outstanding 2 conversations, I'm fine with that. I'm happy to fix MiMa and make changes based on the resolutions of the conversations, but I don't know how soon that would happen, so if someone else wants to, that's fine with me

**Comment by SethTisue** [comment]
anyone interested in pursuing this further...?

**Comment by NthPortal** [comment]
I'll try to get to it... this weekend maybe? working again, so I don't have a lot of capacity these days

**Comment by NthPortal** [comment]
@SethTisue I've addressed all the comments (and added MiMa filters). unless anyone else wants to review, I think it's good to go once the build finishes

**Comment by lrytz** [comment]
LGTM, but it would be good if another pair of eyes with brain behind could take a look.

File: src/library/scala/collection/mutable/ArrayBuffer.scala
Comment (SethTisue) [inline src/library/scala/collection/mutable/ArrayBuffer.scala]: oops, this is out of sync with the code (`private` vs `protected`)

**Review by SethTisue — DISMISSED** [review]

**Comment by SethTisue** [comment]
Consider this approved for merge once my one tiny nitpick is fixed.

File: src/library/scala/collection/mutable/ArrayBuffer.scala
Comment (NthPortal) [reply src/library/scala/collection/mutable/ArrayBuffer.scala]: @SethTisue that was actually intentional. `protected[x]` is the union of `protected` and `private[x]`. to fulfill the needs of the stdlib (`PriorityQueue` in particular), it must be `protected`. parallel-collections requires it to be `private[collection]`, but does not make use of it being `protected`.

Comment (SethTisue) [reply src/library/scala/collection/mutable/ArrayBuffer.scala]: Okay — but the comment doesn't communicate that to the reader. As written, it _reads_ like a mistake.

Comment (NthPortal) [reply src/library/scala/collection/mutable/ArrayBuffer.scala]: ... stop making good points Seth

**Pull request merged**
